File: crates/worldgen/src/geo.rs
//! Port of Valheim's `WorldGenerator` terrain + biome classification.
//!
//! Transcribed from the decompiled `WorldGenerator.cs` (cross-checked across
//! six independent mirrors and the MIT-licensed Avledet C++ port). The core
//! math below was byte-stable from game build 0.216.5 through 0.221.12.

use crate::perlin::perlin;
use crate::random::{seed_from_name, UnityRandom};
use crate::fastnoise::FastNoise;
use crate::rivers::Rivers;

pub const WORLD_SIZE: f32 = 10000.0;
pub const WATER_EDGE: f32 = 10500.0;
pub const WATER_TRESHOLD: f32 = 0.05;
pub const HEIGHT_MULTIPLIER: f32 = 200.0;
pub const ASHLANDS_MIN_DISTANCE: f32 = 12000.0;
pub const ASHLANDS_Y_OFFSET: f32 = -4000.0;
pub const DEEP_NORTH_MIN_DISTANCE: f32 = 12000.0;
pub const DEEP_NORTH_Y_OFFSET: f32 = 4000.0;
/// Radii outside which the polar ocean gaps are provably 1.0 — the arc radius
/// (12000 ± 100 from the angle term) widened by the 400 m gap falloff.
const GAP_INNER: f32 = 11400.0;
const GAP_OUTER: f32 = 12600.0;
/// Sea level in world Y units.
pub const WATER_LEVEL: f32 = 30.0;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[repr(u16)]
pub enum Biome {
    None = 0,
    Meadows = 1,
    Swamp = 2,
    Mountain = 4,
    BlackForest = 8,
    Plains = 16,
    AshLands = 32,
    DeepNorth = 64,
    Ocean = 256,
    Mistlands = 512,
}

/// `DUtils.PerlinNoise(double, double)` — the coordinates are computed wide
/// and narrowed only at the call boundary.
#[inline(always)]
pub fn pn(x: f64, y: f64) -> f64 {
    perlin(x as f32, y as f32) as f64
}

#[inline(always)]
pub fn length(x: f32, y: f32) -> f32 {
    // `DUtils.Length` squares and sums in double before the sqrt. In f32 the
    // sum reaches ~2e8 at world-edge distances, where the representable step
    // is ~16, so the radius comes back wrong by up to ~4e-4 m — enough to sit
    // on the wrong side of the `dist > 2000` / `dist < 6000` biome-radius
    // comparisons for a handful of pixels.
    (((x as f64) * (x as f64) + (y as f64) * (y as f64)).sqrt()) as f32
}

/// `DUtils.Lerp`. Two details that look like noise and are not: it clamps at
/// both ends, and it interpolates as `a*(1-t) + b*t` in double rather than the
/// usual `a + (b-a)*t`. The two forms disagree in the last bits.
#[inline(always)]
pub fn lerp(a: f32, b: f32, t: f32) -> f32 {
    if t <= 0.0 { return a; }
    if t >= 1.0 { return b; }
    ((a as f64) * (1.0 - t as f64) + (b as f64) * (t as f64)) as f32
}

#[inline(always)]
pub fn lerp_f64(a: f64, b: f64, t: f64) -> f64 {
    if t <= 0.0 { return a; }
    if t >= 1.0 { return b; }
    a * (1.0 - t) + b * t
}

/// `DUtils.Fbm` — the **double** overload. The f32 one in `heights.rs` is
/// the separate float overload the forest factor calls; the game has both.
#[inline(always)]
pub fn fbm_f64(px: f32, py: f32, octaves: i32, lacunarity: f64, gain: f64) -> f64 {
    let mut sum = 0.0;
    let mut amp = 1.0;
    let mut x = px as f64;
    let mut y = py as f64;
    for _ in 0..octaves {
        sum += amp * pn(x, y);
        amp *= gain;
        x *= lacunarity;
        y *= lacunarity;
    }
    sum
}

#[inline(always)]
pub fn clamp01(v: f32) -> f32 {
    if v > 1.0 { 1.0 } else if v < 0.0 { 0.0 } else { v }
}

#[inline(always)]
pub fn lerp_step(l: f32, h: f32, v: f32) -> f32 {
    clamp01((v - l) / (h - l))
}

#[inline(always)]
pub fn smooth_step(lo: f32, hi: f32, x: f32) -> f32 {
    let t = clamp01((((x as f64) - lo as f64) / ((hi as f64) - lo as f64)) as f32) as f64;
    (t * t * (3.0 - 2.0 * t)) as f32
}

#[inline(always)]
pub fn clamp01_f64(v: f64) -> f64 {
    if v > 1.0 { 1.0 } else if v < 0.0 { 0.0 } else { v }
}

/// `DUtils.MathfLikeSmoothStep`. The game narrows the result to f32 before
/// widening it back to f64 on return; that rounding is load-bearing for the
/// ocean-gap edge, so it is reproduced rather than tidied away.
#[inline(always)]
pub fn mathf_like_smooth_step(from: f64, to: f64, t: f64) -> f64 {
    let t = clamp01_f64(t);
    let t = -2.0 * t * t * t + 3.0 * t * t;
    ((to * t + from * (1.0 - t)) as f32) as f64
}

#[inline(always)]
pub fn blend_overlay(a: f64, b: f64) -> f64 {
    if a < 0.5 { 2.0 * a * b } else { 1.0 - 2.0 * (1.0 - a) * (1.0 - b) }
}

#[inline(always)]
pub fn remap(value: f64, in_lo: f64, in_hi: f64, out_lo: f64, out_hi: f64) -> f64 {
    let t = if in_lo == in_hi { 0.0 } else { clamp01_f64((value - in_lo) / (in_hi - in_lo)) };
    out_lo + (out_hi - out_lo) * t
}

pub struct WorldGenerator {
    pub seed: i32,
    pub offset0: f32,
    pub offset1: f32,
    pub offset2: f32,
    pub offset3: f32,
    pub offset4: f32,
    pub river_seed: i32,
    pub stream_seed: i32,

    // VersionSetup(): tunables that differ for legacy worlds.
    min_mountain_distance: f32,
    max_marsh_distance: f32,
    min_darkland_noise: f32,

    pub rivers: Rivers,
    /// Cellular/simplex source for Ashlands terrain. The game holds this in a
    /// `static` guarded by a null check, so in-game it keeps the seed of the
    /// first world loaded that session; one instance per generator reproduces
    /// the intended (and overwhelmingly common) single-world case.
    pub noise: FastNoise,
}

impl WorldGenerator {
    pub fn from_seed_name(name: &str, world_gen_version: i32) -> Self {
        Self::from_seed(seed_from_name(name), world_gen_version)
    }

    /// A generator with the offsets drawn but **no river/lake/stream
    /// pregeneration**. Biome classification never reads river data — verified
    /// bit-identical over 125 627 samples — so seed search can screen
    /// candidates on biomes alone for ~1 ms instead of ~145 ms. Heights from
    /// this instance are wrong wherever water carved the terrain; use it only
    /// for `biome`/`base_height`.
    pub fn biome_only(seed: i32, world_gen_version: i32) -> Self {
        Self::build(seed, world_gen_version, false)
    }

    pub fn from_seed(seed: i32, world_gen_version: i32) -> Self {
        Self::build(seed, world_gen_version, true)
    }

    fn build(seed: i32, world_gen_version: i32, pregenerate: bool) -> Self {
        // VersionSetup(version)
        let mut min_mountain_distance = 1000.0f32;
        let mut max_marsh_distance = 6000.0f32;
        let mut min_darkland_noise = 0.4f32;
        if world_gen_version <= 0 {
            min_mountain_distance = 1500.0;
        }
        if world_gen_version <= 1 {
            min_darkland_noise = 0.5;
            max_marsh_distance = 8000.0;
        }

        // Draw order is load-bearing: offset0..3, riverSeed, streamSeed, then
        // offset4 LAST. Getting this wrong silently produces a different world.
        let mut r = UnityRandom::new(seed);
        let offset0 = r.range_i32(-10000, 10000) as f32;
        let offset1 = r.range_i32(-10000, 10000) as f32;
        let offset2 = r.range_i32(-10000, 10000) as f32;
        let offset3 = r.range_i32(-10000, 10000) as f32;
        let river_seed = r.range_i32(i32::MIN, i32::MAX);
        let stream_seed = r.range_i32(i32::MIN, i32::MAX);
        let offset4 = r.range_i32(-10000, 10000) as f32;

        let mut wg = Self {
            seed,
            offset0,
            offset1,
            offset2,
            offset3,
            offset4,
            river_seed,
            stream_seed,
            min_mountain_distance,
            max_marsh_distance,
            min_darkland_noise,
            rivers: Rivers::empty(),
            noise: FastNoise::new(seed),
        };
        if pregenerate {
            wg.pregenerate();
        }
        wg
    }

    #[inline(always)]
    pub fn world_angle(&self, wx: f32, wy: f32) -> f32 {
        // NOTE: atan2(x, y), not the conventional atan2(y, x).
        (wx.atan2(wy) * 20.0).sin()
    }

    /// `GetBaseHeight`. This is the one generator function that keeps its
    /// coordinate accumulation in **double** all the way to the noise call —
    /// `double num5 = wx; num5 += 100000.0 + (double)m_offset0;` with no
    /// intermediate narrowing. Every per-biome height function does the
    /// opposite (`wx = (float)((double)wx + 100000.0 + m_offset3)`), so only
    /// this one needs the wider type.
    ///
    /// Running it in f32 throughout is visually indistinguishable — measured
    /// max divergence 5.0e-5, i.e. 1 cm of terrain — but it flips roughly one
    /// pixel in 400 000 across the ocean/mountain thresholds, and those flips
    /// land on coastlines where they are most visible.
    pub fn base_height(&self, wx_in: f32, wy_in: f32) -> f32 {
        let dist = length(wx_in, wy_in);
        let x = wx_in as f64 + 100000.0 + self.offset0 as f64;
        let y = wy_in as f64 + 100000.0 + self.offset1 as f64;

        // The literals are the f32 constants the game uses, written out at the
        // width the double-precision path actually sees them.
        const A: f64 = 0.0020000000949949026;
        const B: f64 = 0.003000000026077032;
        const C: f64 = 0.004999999888241291;
        const D: f64 = 0.009999999776482582;

        let mut h = 0.0f32;
        h = (h as f64 + pn(x * A * 0.5, y * A * 0.5) * pn(x * B * 0.5, y * B * 0.5)) as f32;
        h = (h as f64 + pn(x * A, y * A) * pn(x * B, y * B) * h as f64 * 0.8999999761581421) as f32;
        h = (h as f64 + pn(x * C, y * C) * pn(x * D, y * D) * 0.5 * h as f64) as f32;
        h = (h as f64 - 0.07000000029802322) as f32;

        // Flattening mask: carves broad lowland plateaus away from the centre.
        let n4 = pn(x * A * 0.25 + 0.12300000339746475, y * A * 0.25 + 0.15123000741004944);
        let n5 = pn(x * A * 0.25 + 0.32100000977516174, y * A * 0.25 + 0.23100000619888306);
        let v = ((n4 - n5) as f32).abs();
        let mut mask = 1.0 - lerp_step(0.02, 0.12, v) as f64;
        mask *= smooth_step(744.0, 1000.0, dist) as f64;
        h = ((h as f64) * (1.0 - mask)) as f32;

        // World edge falloff.
        if dist > WORLD_SIZE {
            let t = lerp_step(WORLD_SIZE, WATER_EDGE, dist);
            h = lerp(h, -0.2, t);
            const EDGE: f32 = 10490.0;
            if dist > EDGE {
                let t2 = lerp_step(EDGE, WATER_EDGE, dist);
                h = lerp(h, -2.0, t2);
            }
            return h;
        }

        // Keep big mountains away from spawn.
        if dist < self.min_mountain_distance && h > 0.28 {
            let t3 = clamp01((((h as f64) - 0.2800000011920929) / 0.09999999403953552) as f32);
            h = lerp(
                lerp(0.28, 0.38, t3),
                h,
                lerp_step(self.min_mountain_distance - 400.0, self.min_mountain_distance, dist),
            );
        }
        h
    }

    /// How far past the Ashlands boundary a point is, in units of 300 m.
    /// Note the angle term is sampled at `wy + ashlandsYOffset` here but at
    /// plain `wy` in `is_ashlands` — that asymmetry is in the game.
    #[inline]
    pub fn ashlands_ocean_gradient(&self, wx: f32, wy: f32) -> f32 {
        let a = self.world_angle(wx, wy + ASHLANDS_Y_OFFSET) as f64 * 100.0;
        ((length(wx, wy + ASHLANDS_Y_OFFSET) as f64 - (ASHLANDS_MIN_DISTANCE as f64 + a)) / 300.0)
            as f32
    }

    /// `CreateAshlandsGap` — a multiplier on the height scale, not on the
    /// height itself. It reaches 0 exactly on the Ashlands biome boundary and
    /// returns to 1 over 400 m either side, which is what drowns the border
    /// into the ocean strip that makes Ashlands an island.
    #[inline]
    pub fn create_ashlands_gap(&self, wx: f32, wy: f32) -> f64 {
        // The angle term is bounded by ±100, so the arc always lies within
        // [11900, 12100] and the gap is only below 1 within 400 m of it.
        // Anything outside [11400, 12600] is unambiguously untouched, which
        // lets the common case skip an atan2 and a sin — worth doing, because
        // this runs on every height sample in the world.
        let r = length(wx, wy + ASHLANDS_Y_OFFSET);
        if !(GAP_INNER..=GAP_OUTER).contains(&r) {
            return 1.0;
        }
        let a = self.world_angle(wx, wy) as f64 * 100.0;
        let v = r as f64 - (ASHLANDS_MIN_DISTANCE as f64 + a);
        let v = clamp01_f64(v.abs() / 400.0) as f32;
        mathf_like_smooth_step(0.0, 1.0, v as f64)
    }

    /// `CreateDeepNorthGap` — the same mechanism at the north pole.
    #[inline]
    pub fn create_deep_north_gap(&self, wx: f32, wy: f32) -> f64 {
        let r = length(wx, wy + DEEP_NORTH_Y_OFFSET);
        if !(GAP_INNER..=GAP_OUTER).contains(&r) {
            return 1.0;
        }
        let a = self.world_angle(wx, wy) as f64 * 100.0;
        let v = r as f64 - (DEEP_NORTH_MIN_DISTANCE as f64 + a);
        let v = clamp01_f64(v.abs() / 400.0) as f32;
        mathf_like_smooth_step(0.0, 1.0, v as f64)
    }

    #[inline]
    pub fn is_ashlands(&self, wx: f32, wy: f32) -> bool {
        let a = self.world_angle(wx, wy) * 100.0;
        length(wx, wy + -4000.0) > 12000.0 + a
    }

    #[inline]
    pub fn is_deep_north(&self, wx: f32, wy: f32) -> bool {
        let a = self.world_angle(wx, wy) * 100.0;
        length(wx, wy + 4000.0) > 12000.0 + a
    }

    #[inline]
    pub fn biome(&self, wx: f32, wy: f32) -> Biome {
        self.biome_with_base(wx, wy).0
    }

    /// Returns the biome alongside the base height it was derived from, so
    /// callers that also need terrain height don't pay for it twice.
    pub fn biome_with_base(&self, wx: f32, wy: f32) -> (Biome, f32) {
        let dist = length(wx, wy);
        let base = self.base_height(wx, wy);
        let a = self.world_angle(wx, wy) * 100.0;

        if length(wx, wy + -4000.0) > 12000.0 + a {
            return (Biome::AshLands, base);
        }
        if base <= 0.02 {
            return (Biome::Ocean, base);
        }
        if length(wx, wy + 4000.0) > 12000.0 + a {
            return if base > 0.4 { (Biome::Mountain, base) } else { (Biome::DeepNorth, base) };
        }
        if base > 0.4 {
            return (Biome::Mountain, base);
        }
        if perlin((self.offset0 + wx) * 0.001, (self.offset0 + wy) * 0.001) > 0.6
            && dist > 2000.0
            && dist < self.max_marsh_distance
            && base > 0.05
            && base < 0.25
        {
            return (Biome::Swamp, base);
        }
        if perlin((self.offset4 + wx) * 0.001, (self.offset4 + wy) * 0.001) > self.min_darkland_noise
            && dist > 6000.0 + a
            && dist < 10000.0
        {
            return (Biome::Mistlands, base);
        }
        if perlin((self.offset1 + wx) * 0.001, (self.offset1 + wy) * 0.001) > 0.4
            && dist > 3000.0 + a
            && dist < 8000.0
        {
            return (Biome::Plains, base);
        }
        if perlin((self.offset2 + wx) * 0.001, (self.offset2 + wy) * 0.001) > 0.4
            && dist > 600.0 + a
            && dist < 6000.0
        {
            return (Biome::BlackForest, base);
        }
        if dist > 5000.0 + a {
            return (Biome::BlackForest, base);
        }
        (Biome::Meadows, base)
    }
}

File: apple/Sources/Worldgen/Geo.swift
//  Geo.swift
//  Port of Valheim's `WorldGenerator` terrain + biome classification.
//  Follows the porting contract in UnityRandom.swift.

import Foundation

public let WORLD_SIZE: Float = 10000.0
public let WATER_EDGE: Float = 10500.0
public let HEIGHT_MULTIPLIER: Float = 200.0
public let ASHLANDS_MIN_DISTANCE: Float = 12000.0
public let ASHLANDS_Y_OFFSET: Float = -4000.0
public let DEEP_NORTH_MIN_DISTANCE: Float = 12000.0
public let DEEP_NORTH_Y_OFFSET: Float = 4000.0
/// Radii outside which the polar ocean gaps are provably 1.0.
private let GAP_INNER: Float = 11400.0
private let GAP_OUTER: Float = 12600.0
/// Sea level in world Y units.
public let WATER_LEVEL: Float = 30.0

public enum Biome: UInt16, Sendable, CaseIterable {
    case none = 0
    case meadows = 1
    case swamp = 2
    case mountain = 4
    case blackForest = 8
    case plains = 16
    case ashLands = 32
    case deepNorth = 64
    case ocean = 256
    case mistlands = 512
}

/// `DUtils.Length` squares and sums in double before the sqrt. In f32 the sum
/// reaches ~2e8 at world-edge distances where the representable step is ~16,
/// enough to land on the wrong side of the biome-radius comparisons.
@inline(__always)
public func length(_ x: Float, _ y: Float) -> Float {
    Float((Double(x) * Double(x) + Double(y) * Double(y)).squareRoot())
}

/// `DUtils.Lerp`. Clamps at both ends and interpolates as `a*(1-t) + b*t` in
/// double rather than `a + (b-a)*t`. The two forms disagree in the last bits.
@inline(__always)
public func lerp(_ a: Float, _ b: Float, _ t: Float) -> Float {
    if t <= 0.0 { return a }
    if t >= 1.0 { return b }
    return Float(Double(a) * (1.0 - Double(t)) + Double(b) * Double(t))
}

@inline(__always)
public func lerpF64(_ a: Double, _ b: Double, _ t: Double) -> Double {
    if t <= 0.0 { return a }
    if t >= 1.0 { return b }
    return a * (1.0 - t) + b * t
}

/// `DUtils.Fbm` — the double overload.
@inline(__always)
public func fbmF64(_ px: Float, _ py: Float, _ octaves: Int32, _ lacunarity: Double, _ gain: Double) -> Double {
    var sum = 0.0
    var amp = 1.0
    var x = Double(px)
    var y = Double(py)
    for _ in 0..<octaves {
        sum += amp * pn(x, y)
        amp *= gain
        x *= lacunarity
        y *= lacunarity
    }
    return sum
}

@inline(__always)
public func clamp01(_ v: Float) -> Float { v > 1.0 ? 1.0 : (v < 0.0 ? 0.0 : v) }

@inline(__always)
public func lerpStep(_ l: Float, _ h: Float, _ v: Float) -> Float { clamp01((v - l) / (h - l)) }

@inline(__always)
public func smoothStep(_ lo: Float, _ hi: Float, _ x: Float) -> Float {
    let t = Double(clamp01(Float((Double(x) - Double(lo)) / (Double(hi) - Double(lo)))))
    return Float(t * t * (3.0 - 2.0 * t))
}

@inline(__always)
public func clamp01F64(_ v: Double) -> Double { v > 1.0 ? 1.0 : (v < 0.0 ? 0.0 : v) }

/// `DUtils.MathfLikeSmoothStep`. The game narrows to f32 before widening back
/// to f64 on return; that rounding is load-bearing for the ocean-gap edge.
@inline(__always)
public func mathfLikeSmoothStep(_ from: Double, _ to: Double, _ t0: Double) -> Double {
    let t = clamp01F64(t0)
    let s = -2.0 * t * t * t + 3.0 * t * t
    return Double(Float(to * s + from * (1.0 - s)))
}

@inline(__always)
public func blendOverlay(_ a: Double, _ b: Double) -> Double {
    a < 0.5 ? 2.0 * a * b : 1.0 - 2.0 * (1.0 - a) * (1.0 - b)
}

@inline(__always)
public func remap(_ value: Double, _ inLo: Double, _ inHi: Double, _ outLo: Double, _ outHi: Double) -> Double {
    let t = inLo == inHi ? 0.0 : clamp01F64((value - inLo) / (inHi - inLo))
    return outLo + (outHi - outLo) * t
}

/// Immutable once `init` returns: every stored property is `let`, and the one
/// reference field (`rivers`) is only written during pregeneration inside that
/// init. Sharing a finished generator across tasks is therefore safe, which the
/// compiler cannot see on its own.
public final class WorldGenerator: @unchecked Sendable {
    public let seed: Int32
    public let offset0: Float
    public let offset1: Float
    public let offset2: Float
    public let offset3: Float
    public let offset4: Float
    public let riverSeed: Int32
    public let streamSeed: Int32

    // VersionSetup(): tunables that differ for legacy worlds.
    let minMountainDistance: Float
    let maxMarshDistance: Float
    let minDarklandNoise: Float

    public var rivers = Rivers()
    /// Seeded **0, not the world seed** — the game constructs it with the world
    /// seed then immediately calls SetSeed(0), so the argument never survives.
    public let noise = FastNoise(seed: 0)

    public convenience init(seedName: String, worldGenVersion: Int32) {
        self.init(seed: seedFromName(seedName), worldGenVersion: worldGenVersion, pregenerate: true)
    }

    /// Offsets drawn but **no river/lake/stream pregeneration**. Biome
    /// classification never reads river data, so seed search can screen on
    /// biomes alone. Heights from this instance are wrong wherever water carved.
    public convenience init(biomeOnly seed: Int32, worldGenVersion: Int32) {
        self.init(seed: seed, worldGenVersion: worldGenVersion, pregenerate: false)
    }

    public init(seed: Int32, worldGenVersion: Int32, pregenerate doPregen: Bool) {
        // VersionSetup(version)
        var minMountain: Float = 1000.0
        var maxMarsh: Float = 6000.0
        var minDarkland: Float = 0.4
        if worldGenVersion <= 0 { minMountain = 1500.0 }
        if worldGenVersion <= 1 { minDarkland = 0.5; maxMarsh = 8000.0 }
        minMountainDistance = minMountain
        maxMarshDistance = maxMarsh
        minDarklandNoise = minDarkland

        // Draw order is load-bearing: offset0..3, riverSeed, streamSeed, then
        // offset4 LAST. Getting this wrong silently produces a different world.
        var r = UnityRandom(seed: seed)
        self.seed = seed
        offset0 = Float(r.range(Int32(-10000), Int32(10000)))
        offset1 = Float(r.range(Int32(-10000), Int32(10000)))
        offset2 = Float(r.range(Int32(-10000), Int32(10000)))
        offset3 = Float(r.range(Int32(-10000), Int32(10000)))
        riverSeed = r.range(Int32.min, Int32.max)
        streamSeed = r.range(Int32.min, Int32.max)
        offset4 = Float(r.range(Int32(-10000), Int32(10000)))

        if doPregen { pregenerate() }
    }

    /// NOTE: atan2(x, y), not the conventional atan2(y, x).
    @inline(__always)
    public func worldAngle(_ wx: Float, _ wy: Float) -> Float {
        mathfSin(mathfAtan2(wx, wy) * 20.0)
    }

    /// `GetBaseHeight`. The one generator function that keeps its coordinate
    /// accumulation in **double** all the way to the noise call. Running it in
    /// f32 throughout flips roughly one pixel in 400 000 across the
    /// ocean/mountain thresholds, and those flips land on coastlines.
    public func baseHeight(_ wxIn: Float, _ wyIn: Float) -> Float {
        let dist = length(wxIn, wyIn)
        let x = Double(wxIn) + 100000.0 + Double(offset0)
        let y = Double(wyIn) + 100000.0 + Double(offset1)

        // The f32 constants the game uses, written at the width the
        // double-precision path actually sees them.
        let A = 0.0020000000949949026
        let B = 0.003000000026077032
        let C = 0.004999999888241291
        let D = 0.009999999776482582

        var h: Float = 0.0
        h = Float(Double(h) + pn(x * A * 0.5, y * A * 0.5) * pn(x * B * 0.5, y * B * 0.5))
        h = Float(Double(h) + pn(x * A, y * A) * pn(x * B, y * B) * Double(h) * 0.8999999761581421)
        h = Float(Double(h) + pn(x * C, y * C) * pn(x * D, y * D) * 0.5 * Double(h))
        h = Float(Double(h) - 0.07000000029802322)

        // Flattening mask: carves broad lowland plateaus away from the centre.
        let n4 = pn(x * A * 0.25 + 0.12300000339746475, y * A * 0.25 + 0.15123000741004944)
        let n5 = pn(x * A * 0.25 + 0.32100000977516174, y * A * 0.25 + 0.23100000619888306)
        let v = abs(Float(n4 - n5))
        var mask = 1.0 - Double(lerpStep(0.02, 0.12, v))
        mask *= Double(smoothStep(744.0, 1000.0, dist))
        h = Float(Double(h) * (1.0 - mask))

        // World edge falloff.
        if dist > WORLD_SIZE {
            let t = lerpStep(WORLD_SIZE, WATER_EDGE, dist)
            h = lerp(h, -0.2, t)
            let EDGE: Float = 10490.0
            if dist > EDGE {
                let t2 = lerpStep(EDGE, WATER_EDGE, dist)
                h = lerp(h, -2.0, t2)
            }
            return h
        }

        // Keep big mountains away from spawn.
        if dist < minMountainDistance && h > 0.28 {
            let t3 = clamp01(Float((Double(h) - 0.2800000011920929) / 0.09999999403953552))
            h = lerp(
                lerp(0.28, 0.38, t3),
                h,
                lerpStep(minMountainDistance - 400.0, minMountainDistance, dist)
            )
        }
        return h
    }

    /// `CreateAshlandsGap` — a multiplier on the height scale. Reaches 0 exactly
    /// on the Ashlands boundary and returns to 1 over 400 m either side.
    @inline(__always)
    public func createAshlandsGap(_ wx: Float, _ wy: Float) -> Double {
        let r = length(wx, wy + ASHLANDS_Y_OFFSET)
        if r < GAP_INNER || r > GAP_OUTER { return 1.0 }
        let a = Double(worldAngle(wx, wy)) * 100.0
        let v = Double(r) - (Double(ASHLANDS_MIN_DISTANCE) + a)
        let c = Float(clamp01F64(abs(v) / 400.0))
        return mathfLikeSmoothStep(0.0, 1.0, Double(c))
    }

    /// `CreateDeepNorthGap` — the same mechanism at the north pole.
    @inline(__always)
    public func createDeepNorthGap(_ wx: Float, _ wy: Float) -> Double {
        let r = length(wx, wy + DEEP_NORTH_Y_OFFSET)
        if r < GAP_INNER || r > GAP_OUTER { return 1.0 }
        let a = Double(worldAngle(wx, wy)) * 100.0
        let v = Double(r) - (Double(DEEP_NORTH_MIN_DISTANCE) + a)
        let c = Float(clamp01F64(abs(v) / 400.0))
        return mathfLikeSmoothStep(0.0, 1.0, Double(c))
    }

    @inline(__always)
    public func biome(_ wx: Float, _ wy: Float) -> Biome { biomeWithBase(wx, wy).0 }

    /// Returns the biome alongside the base height it was derived from, so
    /// callers that also need terrain height don't pay for it twice.
    public func biomeWithBase(_ wx: Float, _ wy: Float) -> (Biome, Float) {
        let dist = length(wx, wy)
        let base = baseHeight(wx, wy)
        let a = worldAngle(wx, wy) * 100.0

        if length(wx, wy + -4000.0) > 12000.0 + a { return (.ashLands, base) }
        if base <= 0.02 { return (.ocean, base) }
        if length(wx, wy + 4000.0) > 12000.0 + a {
            return base > 0.4 ? (.mountain, base) : (.deepNorth, base)
        }
        if base > 0.4 { return (.mountain, base) }
        if perlin((offset0 + wx) * 0.001, (offset0 + wy) * 0.001) > 0.6
            && dist > 2000.0 && dist < maxMarshDistance
            && base > 0.05 && base < 0.25 {
            return (.swamp, base)
        }
        if perlin((offset4 + wx) * 0.001, (offset4 + wy) * 0.001) > minDarklandNoise
            && dist > 6000.0 + a && dist < 10000.0 {
            return (.mistlands, base)
        }
        if perlin((offset1 + wx) * 0.001, (offset1 + wy) * 0.001) > 0.4
            && dist > 3000.0 + a && dist < 8000.0 {
            return (.plains, base)
        }
        if perlin((offset2 + wx) * 0.001, (offset2 + wy) * 0.001) > 0.4
            && dist > 600.0 + a && dist < 6000.0 {
            return (.blackForest, base)
        }
        if dist > 5000.0 + a { return (.blackForest, base) }
        return (.meadows, base)
    }
}
